### PR TITLE
Browse zip archives & other things

### PR DESCRIPTION
**Pull request by gave92**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- For #45
- For #5833
- For #4180

**Details of Changes**
Add details of changes here.
Replaced most of references to StorageFile and StorageFolder with new base classes that implements IStorageItem (BaseStorageFile[Folder]). The standard StorageFile[Folder], the FtpStorageFile[Folder] and the new ZipStorageFile[Folder] all derive from these base classes. This enables the following features "for free".

**Features enabled by this PR**
- Browse zip archives as folders
- Add files to zip by drag & drop
- Delete from zip
- Open FTP files with double click
- Search and navigation bar suggestions in FTP folders
- Copy folders to/from FTP
- Preview files in zip and ftp folders

Comments, testing and contributions are welcome xD

**Validation**
How did you test these changes?
- [x] Built and ran the app

### CODE REVIEW

**Comment by winston-de** [comment]
@gave92 do you mind if I implement making files a handler for zip files to the pr (so that it can be set as the default app for archives)?

File: Files/ViewModels/SettingsViewModel.cs [hunk 1]
```diff
@@ -81,6 +81,7 @@ private void TrackAnalytics()
             Analytics.TrackEvent($"{nameof(ShowLibrarySection)} {ShowLibrarySection}");
             Analytics.TrackEvent($"{nameof(ShowBundlesWidget)} {ShowBundlesWidget}");
             Analytics.TrackEvent($"{nameof(ListAndSortDirectoriesAlongsideFiles)} {ListAndSortDirectoriesAlongsideFiles}");
+            Analytics.TrackEvent($"{nameof(AreFileTagsEnabled)} {AreFileTagsEnabled}");
```
Comment (yaichenbaum) [inline Files/ViewModels/SettingsViewModel.cs:84]: Should this be changed to `AreTagsEnabled` since tags work for folders as well?

**Comment by gave92** [comment]
> @gave92 do you mind if I implement making files a handler for zip files to the pr (so that it can be set as the default app for archives)?

That's a great idea, please do 👍

File: Files/ViewModels/SettingsViewModel.cs (hunk 1 above)
Comment (gave92) [reply Files/ViewModels/SettingsViewModel.cs:84]: I'd leave the variables name as they are, if any we could change the user-facing string resources to something different than "file tags", but just "tags" doesn't sound clear 🤔

Comment (yaichenbaum) [reply Files/ViewModels/SettingsViewModel.cs:84]: I'm not sure file tags is much clearer 🤔. Perhaps we should add information to the docs or a teaching tip to the app.

**Comment by winston-de** [comment]
Love it, just a couple of minor things I noticed:

- dragging a file from a zip to another directory moves it, should it copy instead?
- changes to zip files don't seem to be detected, so operations like delete don't show until refreshing

**Comment by gave92** [comment]
> * dragging a file from a zip to another directory moves it, should it copy instead?

True, the default action should be copy. There's also another issue for which dragging a file _into_ a zip archive defaults to move (which fails), default should be copy here as well.

> * changes to zip files don't seem to be detected, so operations like delete don't show until refreshing

True, it's the same for FTP, I don't think I'm adding either here

**Comment by d2dyno1** [comment]
Wow! Excellent job @gave92!

**Comment by winston-de** [comment]
I implemented making copy the default operation when dragging a file from zip over a tab by using the same function for dropping into the layout blank space (also reduces code duplication), is that good with you?

**Comment by gave92** [comment]
I hate CodeMaid lol

**Review by winston-de — APPROVED** [review]

**Comment by winston-de** [comment]
Amazing

File: Files/Filesystem/StorageItems/FtpStorageFile.cs [hunk 2]
```diff
@@ -8,47 +8,59 @@
...
 
         private async void FtpDataStreamingHandler(StreamedFileDataRequest request)
         {
             try
             {
-                var ftpClient = _viewModel.GetFtpInstance();
+                using var ftpClient = new FtpClient();
+                ftpClient.Host = FtpHelpers.GetFtpHost(Path);
+                ftpClient.Port = FtpHelpers.GetFtpPort(Path);
+                ftpClient.Credentials = FtpManager.Credentials.Get(ftpClient.Host, FtpManager.Anonymous);
```
Comment (hez2010) [inline Files/Filesystem/StorageItems/FtpStorageFile.cs:51]: I think this should also be distinguished by port.

File: Files/ViewModels/Previews/TextPreviewViewModel.cs [hunk 3]
```diff
@@ -44,8 +44,8 @@ public static async Task<TextPreview> TryLoadAsTextAsync(ListedItem item)
 
             try
             {
-                item.ItemFile = await StorageFile.GetFileFromPathAsync(item.ItemPath);
-                var text = await FileIO.ReadTextAsync(item.ItemFile);
+                item.ItemFile = await StorageFileExtensions.DangerousGetFileFromPathAsync(item.ItemPath);
+                var text = await ReadFileAsText(item.ItemFile); // await FileIO.ReadTextAsync(item.ItemFile);
```
[the remark this answers is not in the document]
Comment (d2dyno1) [reply Files/ViewModels/Previews/TextPreviewViewModel.cs:48]: Also while at it, `StorageFileExtensions.DangerousGetFileFromPathAsync` could be replaced with safer `StorageItemHelpers.ToStorageItem<StorageFile>(item.Path);`

[the remark this answers is not in the document]
Comment (gave92) [reply Files/ViewModels/Previews/TextPreviewViewModel.cs:48]: @winston-de `FileIO.ReadTextAsync` did not work for some reason on FTP but if it causes issues with normal previews and can't be easily fixed we can live without previews on FTP paths.
@d2dyno1 good point. (Note: it's `StorageItemHelpers.ToStorageItem<BaseStorageFile>(...)` now)

[the remark this answers is not in the document]
Comment (gave92) [reply Files/ViewModels/Previews/TextPreviewViewModel.cs:48]: Ah well just seen why `ReadFileAsText` always contains "\0\0\0\0", uops.

**Pull request merged**